### PR TITLE
Fix seq wrap of nack, buffer, timestamp wrap of buffer.

### PR DESCRIPTION
**Pull request by cnderrauber**
Fixed:

1. buffer can not push correct nack sequence when sequence wrap. when rtp sequence 65533, 1, 2 received, it should request nack for packet 65534, 65535, 0. @boks1971 comment this problem.
use a SeqWrapHandler to handle sequence wrap, and return the uint32 extend sequence, wrap cycles, etc...

2. nack can not generate nack pair when sequence wrap, and packet 0 is lost. when extend sequence  of lost packets {65536, 65537} push to nack, it should produce nackpair {packetid: 0, bitmap: 1}.

3. IsLaterTimestampWrapAround's comment is not consistence with it's code, and  use 0xc000000 as the wrap check breakpoint

Add some test case for these issue, most of them are boundary conditions.

### CODE REVIEW

File: pkg/buffer/buffer.go [hunk 1]
```diff
@@ -671,19 +650,68 @@ func (b *Buffer) GetLatestTimestamp() (latestTimestamp uint32, latestTimestampTi
...
 func IsTimestampWrapAround(timestamp1 uint32, timestamp2 uint32) bool {
-	return (timestamp1&0xC000000 == 0) && (timestamp2&0xC000000 == 0xC000000)
+	return timestamp2 < timestamp1 && timestamp1 > 0xf0000000 && timestamp2 < 0x0fffffff
 }
 
 // IsLaterTimestamp returns true if timestamp1 is later in time than timestamp2 factoring in timestamp wrap-around
 func IsLaterTimestamp(timestamp1 uint32, timestamp2 uint32) bool {
 	if timestamp1 > timestamp2 {
-		if IsTimestampWrapAround(timestamp2, timestamp1) {
+		if IsTimestampWrapAround(timestamp1, timestamp2) {
```
Comment (davidzhao) [inline pkg/buffer/buffer.go:659]: good catch!

**Review by davidzhao — APPROVED** [review]
solid work! the added tests are great to see too.

from what I can understand, it lg. but will leave to Raja for the final ✅

**Review by boks1971 — APPROVED** [review]
Left a couple of comments, but unable to review on phone much. Sorry. Will look at it more when I can get on the laptop when power comes back. But please go ahead.

File: pkg/buffer/buffer.go [hunk 2]
```diff
@@ -671,19 +650,68 @@ func (b *Buffer) GetLatestTimestamp() (latestTimestamp uint32, latestTimestampTi
 
 // IsTimestampWrapAround returns true if wrap around happens from timestamp1 to timestamp2
 func IsTimestampWrapAround(timestamp1 uint32, timestamp2 uint32) bool {
-	return (timestamp1&0xC000000 == 0) && (timestamp2&0xC000000 == 0xC000000)
+	return timestamp2 < timestamp1 && timestamp1 > 0xf0000000 && timestamp2 < 0x0fffffff
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/buffer/buffer.go:653]: Yes, in beginning I  use the half of range as breakpoint . Then I check the code in webrtc about the timestamp wrap, it has two way to implement, one is half range, another is code commited. Both is ok, use this one just think is easy to understand for other people.

File: pkg/buffer/buffer.go [hunk 3]
```diff
@@ -671,19 +650,68 @@ func (b *Buffer) GetLatestTimestamp() (latestTimestamp uint32, latestTimestampTi
...
 }
 
 // IsLaterTimestamp returns true if timestamp1 is later in time than timestamp2 factoring in timestamp wrap-around
 func IsLaterTimestamp(timestamp1 uint32, timestamp2 uint32) bool {
 	if timestamp1 > timestamp2 {
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/buffer/buffer.go:679]: In case of timestamp wrap, 0xff000000 is earlier thab 0x10

File: pkg/buffer/buffer.go [hunk 4]
```diff
@@ -671,19 +650,68 @@ func (b *Buffer) GetLatestTimestamp() (latestTimestamp uint32, latestTimestampTi
...
 	}
 	return false
 }
+
+func isNewerUint16(val1, val2 uint16) bool {
+	return val1 != val2 && val1-val2 < 0x8000
+}
+
+type SeqWrapHandler struct {
+	// cycles   uint32
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/buffer/buffer.go]: Done.

File: pkg/buffer/buffer.go [hunk 5]
```diff
@@ -75,11 +75,11 @@ type Buffer struct {
 	lastSRRTPTime      uint32
 	lastSRRecv         int64 // Represents wall clock of the most recent sender report arrival
 	baseSN             uint16
-	cycles             uint32
 	lastRtcpPacketTime int64 // Time the last RTCP packet was received.
 	lastRtcpSrTime     int64 // Time the last RTCP SR was received. Required for DLSR computation.
 	lastTransit        uint32
-	maxSeqNo           uint16 // The highest sequence number received in an RTP data packet
+	// maxSeqNo           uint16 // The highest sequence number received in an RTP data packet
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/buffer/buffer.go]: Done

File: pkg/buffer/buffer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/buffer/buffer.go:652]: Cool. Would be good to make utility functions for all the things that need these (sequence number, time stamp, VP8 picture id are the ones that I know of)

File: pkg/buffer/buffer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/buffer/buffer.go:679]: Sorry, my bad. Was confusing things with out-of-order check.

File: pkg/buffer/buffer.go (hunk 1 above)
Comment (boks1971) [reply pkg/buffer/buffer.go:658]: good one!

File: pkg/buffer/buffer.go [hunk 6]
```diff
@@ -341,8 +319,8 @@ func (b *Buffer) calc(pkt []byte, arrivalTime int64) {
 	b.stats.PacketCount++
 
 	ep := ExtPacket{
-		Head:    sn == b.maxSeqNo,
-		Cycle:   b.cycles,
+		Head:    headPkt,
+		Cycle:   b.seqHdlr.Cycles(),
```
Comment (boks1971) [inline pkg/buffer/buffer.go:323]: Looks like this field is not used anywhere. Would be good to get rid of it.

Comment (cnderrauber) [reply pkg/buffer/buffer.go:323]: Agree with this. I will check the code when separate code from ion-sfu, should delete these fields if unused.

File: pkg/buffer/buffer_test.go [hunk 7]
```diff
@@ -53,55 +53,126 @@ func TestNack(t *testing.T) {
...
-			case *rtcp.PictureLossIndication:
-				if p.MediaSSRC == 123 {
-					wg.Done()
+
+	t.Run("nack normal", func(t *testing.T) {
+		buff := NewBuffer(123, pool, pool, logger)
+		buff.codecType = webrtc.RTPCodecTypeVideo
+		assert.NotNil(t, buff)
+		var wg sync.WaitGroup
+		// 3 nacks 1 Pli
```
Comment (boks1971) [inline pkg/buffer/buffer_test.go:62]: Curiosity question: how does this three NACKs happen? Looks like we are dropping one sequence number (1) in the loop and checking that the NACK has that sequence number. How does it happen three times?

File: pkg/buffer/buffer_test.go [hunk 8]
```diff
@@ -53,55 +53,126 @@ func TestNack(t *testing.T) {
...
+		buff.codecType = webrtc.RTPCodecTypeVideo
+		assert.NotNil(t, buff)
+		var wg sync.WaitGroup
+		expects := map[uint16]int{
+			65534: 0,
+			65535: 0,
+			0:     0,
+			1:     0,
+		}
+		wg.Add(3 * len(expects)) // retry 3 times
```
Comment (boks1971) [inline pkg/buffer/buffer_test.go:121]: Looks like three retries on NACK. Not able to see where that is set.

File: pkg/buffer/buffer_test.go [hunk 9]
```diff
@@ -53,55 +53,126 @@ func TestNack(t *testing.T) {
...
+								} else {
+									assert.Fail(t, "unexpected nack seq ", seq)
+								}
+								return true
+							})
+						}
+					}
+				case *rtcp.PictureLossIndication:
+					if p.MediaSSRC == 123 {
+						// wg.Done()
```
Comment (boks1971) [inline pkg/buffer/buffer_test.go:140]: why no PLI in this case? Also, if not expecting PLI, would recommending just deleting this code. Or maybe an assert at the end of the test that no PLIs were received?

File: pkg/buffer/buffer_test.go [hunk 10]
```diff
@@ -236,3 +307,57 @@ func TestFractionLostReport(t *testing.T) {
...
+	}
+	// test normal case, name -> {seq1, seq2, unwrap of seq2}
+	cases := map[string]caseInfo{
+		"no wrap":                             {[]uint32{1, 4, 4}, true},
+		"no wrap backward":                    {[]uint32{4, 1, 1}, false},
+		"wrap around forward to zero":         {[]uint32{65534, 0, 65536}, true},
+		"wrap around forward":                 {[]uint32{65534, 10, 65546}, true},
+		"wrap around forward 2":               {[]uint32{65535 + 65536*2, 1, 1 + 65536*3}, true},
+		"wrap around backward ":               {[]uint32{5, 65534, 65534}, false},
+		"wrap around backward less than zero": {[]uint32{5, 65534, 65534}, false},
```
Comment (boks1971) [inline pkg/buffer/buffer_test.go:328]: @cnderrauber Did you mean to use a different test case here? It is the same as the previous case in the data. Looks like all the cases are covered. Maybe, the last one is not necessary?

File: pkg/buffer/buffer_test.go (hunk 7 above)
Comment (boks1971) [reply pkg/buffer/buffer_test.go:62]: Answering my own question :-) See that in `nacks.go`.

File: pkg/buffer/buffer_test.go (hunk 8 above)
Comment (boks1971) [reply pkg/buffer/buffer_test.go:121]: See the three retries in `nacks.go`

File: pkg/buffer/buffer_test.go (hunk 9 above)
Comment (boks1971) [reply pkg/buffer/buffer_test.go:140]: guess the three retry rule does not trigger the askKF. Okay, some combination of things which does not cause a PLI condition of `ion-sfu`.

**Review by boks1971 — COMMENTED** [review]
Excellent stuff @cnderrauber !!!

File: pkg/buffer/buffer_test.go (hunk 10 above)
Comment (cnderrauber) [reply pkg/buffer/buffer_test.go:328]: Oh, my mistake. I want to add a test case for rtp packets: {65534, 5, 65535}, the sequence wrap forward, then wrap backward. It is difference with the case { 5, 65534 }. In my SeqWrapHandler, has code(buffer.go:700) to handle this, sequence must not unwrap to a negative number:
```
		if delta > 0 && (int32(s.maxSeqNo)+delta-0x10000) >= 0 {
			// wrap backwards, should not less than 0 in this case:
			//   at start time, received seq 1, set s.maxSeqNo =1 ,
			//   then a out of order seq 65534 coming, we can't unwrap
			//   the seq to -2
			delta -= 0x10000
```
Will change the case "wrap around backward" in next pr.

File: pkg/buffer/buffer_test.go (hunk 9 above)
Comment (cnderrauber) [reply pkg/buffer/buffer_test.go:140]: Code of this case is copy from the case before. I use this case  to test nack request for seq wrap, so comment the pli request. Should delete it in next pr.
And for the above questions about nack in ion-sfu, after read the code, I think there are some areas to improve, the interval of nack request generate, retry, timeout, etc... 
But this pr is about the sequence and timestamp fix, so just fix the bug of nack in case of seq wrap.
I have added a todo task of improve nack module in linear, will have a time to work on it.

Comment (cnderrauber) [reply pkg/buffer/buffer_test.go:140]: If you have any suggestion about improve nack module, tell me please!

File: pkg/buffer/buffer_test.go (hunk 10 above)
Comment (boks1971) [reply pkg/buffer/buffer_test.go:328]: Sounds good.

File: pkg/buffer/buffer_test.go (hunk 9 above)
Comment (boks1971) [reply pkg/buffer/buffer_test.go:140]: What you mentioned about RTT is good to implement.

A couple of other things worth considering
1. When a key frame packet is received, clear the NACK queue of all sequence numbers older than that.
2. Should we let DownTrack drive PLI always? Currently, there is some logic for `askKF` in nack.go. Is it possible to let the subscriber always drive it? There is a PLI throttle. So, it is possibly okay to trigger it from multiple places and let the throttle handle it. Was just thinking about simplifying the NACK module. But, if we are doing something like recording without any subscriber, then we have to do the PLI purely in the receiver path.
